frontend/catalyst/jax_tracer.py

```diff
@@ -932,3 +945,124 @@ def is_leaf(obj):
+
+
+    **Example**
+
+    Consider the following circuit:
+
+    .. code-block:: python
+
+        dev = qml.device("lightning.qubit", shots=100)
+        params = np.pi / 4 * np.ones(2)
+
+        @qjit
+        @dynamic_one_shot
+        @qml.qnode(dev, diff_method=None)
+        def circuit(x, y):
+            qml.RX(x, wires=0)
+            m0 = measure(0, reset=reset, postselect=postselect)
+
+            @cond(m0 == 1)
+            def ansatz():
+                qml.RY(y, wires=1)
+
+            ansatz()
+            return measure_f(wires=[0, 1])
+
+    The ``dynamic_one_shot`` decorator prompts the QNode to perform a hundred one-shot
+    calculations, where in each calculation the ``measure`` operations dynamically
+    measures the 0-wire and collapse the state vector stochastically.
+    """
+
+    cpy_tape = None
+    aux_tapes = None
+
+    def transform_to_single_shot(qnode):
+        nonlocal cpy_tape
+        nonlocal aux_tapes
+        cpy_tape = None
+        aux_tapes = None
+
+        @qml.transform
+        def dynamic_one_shot_partial(
+            tape: qml.tape.QuantumTape, **kwargs
+        ) -> (Sequence[qml.tape.QuantumTape], Callable):
+
+            if not any(is_mcm(o) for o in tape.operations):
+                return (tape,), null_postprocessing
+
+            for m in tape.measurements:
+                if not isinstance(
+                    m, (CountsMP, ExpectationMP, ProbabilityMP, SampleMP, VarianceMP)
+                ):
+                    raise TypeError(
+                        f"Native mid-circuit measurement mode does not support {type(m).__name__} "
+                        "measurements."
+                    )
+            _ = kwargs.get("device", None)
+
+            if not tape.shots:
+                raise qml.QuantumFunctionError(
+                    "dynamic_one_shot is only supported with finite shots."
+                )
+
+            samples_present = any(isinstance(mp, SampleMP) for mp in tape.measurements)
+            postselect_present = any(
+                op.postselect is not None for op in tape.operations if is_mcm(op)
+            )
+            if postselect_present and samples_present and tape.batch_size is not None:
+                raise ValueError(
+                    "Returning qml.sample is not supported when postselecting mid-circuit "
+                    "measurements with broadcasting"
+                )
+
+            nonlocal cpy_tape
+            cpy_tape = tape
+            nonlocal aux_tapes
+            aux_tapes = [init_auxiliary_tape(tape)]
+
+            def processing_fn(results):
+                return results
+
+            return aux_tapes, processing_fn
+
+        return dynamic_one_shot_partial(qnode)
+
+    aux_qnode = transform_to_single_shot(qnode)
+    shots = qnode.device.shots
+    single_shot_qnode = aux_qnode
+    single_shot_qnode.device._shots = qml.measurements.Shots(1)
```

Comment:
Would this overwrite the shots on the device if the device instance is used in multiple qnodes?

Reply:
Yes, so I guess we should write shots back on exit.

Reply:
Is there any chance a different qnode could be executed in the meantime though with the wrong shots parameter? 

Should Catalyst instead start looking at the tape for shots?

Reply:
> Is there any chance a different qnode could be executed in the meantime though with the wrong shots parameter?

It could happen with thread-level parallelism, but I don't think this is a supported feature, right?

Reply:
> Should Catalyst instead start looking at the tape for shots?

We could make this a runtime parameter, but I would say this is beyond the scope of the current PR.

Reply:
I didn't mean make it runtime parameter, just reading the value of the tape instead of the device. If I understand correctly, this is what pennylane devices already do?

> It could happen with thread-level parallelism, but I don't think this is a supported feature, right?

Yeah that is probably not something we need to consider, but what about this scenario?
```py
dev = qml.device('lightning.qubit', wires=2, shots=100)

@qjit
@dynamic_one_shot
@qml.qnode(dev)
def main():
   x = jnp.array(5)
   y = inner(x)
   qml.RY(y, wires=0)
   return qml.probs()

@dynamic_one_shot
@qml.qnode(dev)
def inner(x):
    qml.RY(x, wires=0)
    return qml.expval(qml.Z(0))
```

Could it be that shots are set to one when tracing the inner qnode?

Reply:
Oh in my approval I forgot about this issue. Have you verified that the above exampled is not bugged @vincentmr ?

Reply:
This does not work at the moment. I have never considered calling a qnode inside a qnode, is that a requirement?

Reply:
Let's say I try 
```
@qml.qnode(dev)
def inner(x):
    qml.RY(x, wires=0)
    return qml.expval(qml.Z(0))

@qjit
@qml.qnode(dev)
def main():
   x = jnp.array(5)
   y = inner(x)
   qml.RY(y, wires=0)
   return qml.probs()
```
this fails with
```
Exception has occurred: RuntimeError       (note: full exception trace is shown but execution is paused at: _run_module_as_main)
[/home/vincent/cat/runtime/lib/capi/RuntimeCAPI.cpp][Line:251][Function:__catalyst__rt__device_init__impl] Error in Catalyst Runtime: Cannot re-initialize an ACTIVE device: Consider using __catalyst__rt__device_release before __catalyst__rt__device_init
```
Is that the correct way to run this without `dynamic_one_shot`?

Reply:
Huh, I thought this worked 👀

Reply:
> I have never considered calling a qnode inside a qnode, is that a requirement?

I'm not sure, maybe certain qml designs would create compute graphs with nested qnodes? 

But I guess since it's not working right now it's not too important, but we are introducing _another_ reason why this wouldn't be working, which could be dangerous in the future 🤔 If there is a way to ensure the transform is safe now that would be preferable in my opinion.

Reply:
This seems to work 
```
@qml.qnode(dev)
def inner(x):
    qml.RY(x, wires=0)
    return qml.sample(wires=0)


@qjit
@catalyst.qfunc.dynamic_one_shot
@qml.qnode(dev)
def main():
    x = jnp.array(5)
    y = catalyst.qfunc.dynamic_one_shot(inner)(x)
    debug.print("Value of y = {y}", y=y)
    qml.RY(y, wires=0)
    m0 = measure(0)
    return qml.var(m0)
```
`y` appears to be a single sample for each execution.

Reply:
That's bad right? 😅 y should have a length equal to the original number of shots, in each shot of the outer qnode?

Reply:
> That's bad right?

In a sense, maybe, but that's actually what I would like it to do here. In other words, I would not expect to add a batch dimension for each nested QNode, but then again I don't have much experience in this kind of usage. What do you think.

Reply:
Well I mean let's forget about the dynamic_one_shot specifics, if you have a qnode with:

```py
@qml.qnode(qml.device("default.qubit", wires=1, shots=10))
def inner_qnode():
    qml.Hadamard(0)
    return qml.sample(wires=0)
```

I would expect the qnode to produce 10 samples, no matter where it is called (that is the contract of a QNode).
And this should include calling it from other QNodes, which we can verify is also the current behaviour in PennyLane:

```py
@qml.qnode(qml.device("default.qubit", wires=1))
def outer_qnode():
    x = inner_qnode()
    print(x)
    qml.RY(sum(x), wires=0)
    return qml.expval(qml.X(0))

Reply:
In PL `inner_qnode` is executed once when `outer_qnode` is called, but the way the tracing is done in Catalyst, `inner_qnode` is evaluated every time `outer_qnode` is called. Even if we fix the `shots` issue, the frameworks would not be evaluating quite the same thing since in Catalyst `sum(x)` would vary from execution to execution but not PL. Am I wrong?

Reply:
I pushed a fix in [a73048e](https://github.com/PennyLaneAI/catalyst/pull/773/commits/a73048efec228181464aecb80b2e22d1a3ced7cd), but the latter issue remains.

Reply:
I also added the `test_dynamic_one_shot_nested_qnodes` test and realized the example only works in debug mode.

Reply:
> In PL `inner_qnode` is executed once when `outer_qnode` is called, but the way the tracing is done in Catalyst, `inner_qnode` is evaluated every time `outer_qnode` is called. Even if we fix the `shots` issue, the frameworks would not be evaluating quite the same thing since in Catalyst `sum(x)` would vary from execution to execution but not PL. Am I wrong?

That's a good point, which stems from the fact that classical functions are not considered part of the "quantum program" and thus are not part of the shot loop, while in the Catalyst they would be 🤔 

This definitely touches on a deeper issue, and can be generalized to the following issue:
```
@qnode
def dynamic_circuit():
    ...
    m = measure()
    y = classical_fun(m)
    RY(y)
    ...
```

If a hardware device had support for such classical functions during execution, then I think it is clear that they need to execute during every shot of the circuit.

Reply:
> I also added the `test_dynamic_one_shot_nested_qnodes` test and realized the example only works in debug mode.

Oh that's concerning 😅 Well don't worry about the test passing, because as you pointed out nested qnodes weren't working well in Catalyst to begin with. What was important to me was just maintaining code safety by not overwriting the original device. You could make a unit test that just confirms this instead of the system-level test.

Reply:
> What was important to me was just maintaining code safety by not overwriting the original device.

Yep, we now create another device with 
```
qml.device(dev.name, wires=dev.wires, shots=single_shot)
```
for the outer execution. Since we're using `single_shot_qnode` internally, I'm not sure how to validate it, but the system-level test with debug shows it works.

Reply:
Good point, I'm not sure either. Maybe the pytest spy can help?

Reply:
Does this [test](https://github.com/PennyLaneAI/catalyst/pull/773/files#diff-e378d00eb450d844931f323a69879e02e1cdc5e89c3c78f14a9032a84e628f08R300) work for you? The issue with spy is I'm not sure what it would tell us beyond `qml.device` has been called since everything is internal to `dynamic_one_shot`.